### DISCUSSION TITLE
Detect output change

### DISCUSSION THREAD

**Discussion by MaximMaximS** [WayfireWM/wayfire#2271] [Q&A]
Hello,
I need to run a command whenever an output connects or disconnects.

I tried this:
```sh
#!/bin/sh

set -e

PREV="start"
while true; do
    CUR=$(wlr-randr)
    if [ "$CUR" != "$PREV" ]; then
        if [ "$PREV" != "start" ]; then
            # CMD
            # ...
        fi
    fi
    PREV=$CUR
done
```

but it results in high CPU usage.

When I add sleep into the loop, it doesn't use a lot of cpu, but when the output disconnects just for a tiny bit, it doesn't get caught by the script.

Any help would be appreciated.

**Comment by ammen99** [comment]
You can use Wayfire's IPC, here is a small script (you need the `ipc` and `ipc-rules` plugins enabled for it to work, and probably to restart Wayfire so that you get the socket env var):

```py
#!/usr/bin/python3

from wayfire_socket import WayfireSocket
import os

addr = os.getenv('WAYFIRE_SOCKET')

events_sock = WayfireSocket(addr)
events_sock.watch(['output-added', 'output-removed'])

while True:
    msg = events_sock.read_message()
    print(msg) # will contain either output-added or output-removed event
```

You need to copy the `wayfire_socket.py` from the `ipc-scripts/` folder. 

As an alternative, there is also a work-in-progress `wayfire` module in PyPI with similar functionality and similar syntax.
(https://github.com/killown/waypy).

Last but not least, you can use the Wayland protocol itself, but that is much harder to get started.

**Reply by MaximMaximS** [reply]
Hello, thanks for the reply. Unfortunately, I forgot to mention that I am using version 0.7.5, because it is on Raspberry Pi. Do you think that there is any way to achieve my goal on this version?

**Reply by ammen99** [reply]
There is no IPC on Wayfire 0.7.5. You can either write a custom plugin, or use the wayland protocol, take a look at the wayland book and the wl_output globals.
